contrib/party/party-2.12/party.1
.TH PARTY 1  "6 March 1988"
.SH NAME
party \- real-time multi-user communication program
.SH SYNOPSIS
.B party
[
.RI # channel
] [
.I option-list
]
.SH DESCRIPTION
.I party
lets you chat in real time with any other users who are currently running
.IR party .
There are several channels on which separate conversations can take place.
When you run it, it puts you at first into "read" mode.
In this mode lines typed by other users are displayed as they are entered,
each with the speaker's login name as a prefix.
Keep in mind that a person does not see comments entered while he is typing
until he is done typing, so his message may be slightly out of date.
Whenever a person enters or leaves the program, a banner is printed.
.LP
Note that
.I party
keeps all messages in a file that is readable to any user that can run
.IR party .  
It is therefore not a good place to hold private conversations.
.LP
In read mode you can also enter any of several one-character commands
which
temporarily stop the display of messages while you do something else.
.I
party
returns to "read" mode after any command
(except, of course, the quit command).
.IP space 1i
Brings up a ">" prompt so you can type a message.
Messages should not be longer than about 250 characters,
or they may get clipped off.
You can return to "read" mode without entering a message
by entering an EOF (control-D) at the ">" prompt.
.IP "#" 1i
Brings up the "#" prompt so you can enter a new channel name.
You normally start in the channel named "party."
Some channels may be permitted to only certain groups of users.
You can return to "read" mode without changing channels
by entering an EOF (control-D) at the "#" prompt.
If no name is entered, a list of channels is printed with your current
channel marked with an arrow.
The ":join" command is equivalent to this.
.IP "!" 1i
Brings up the "!" prompt so you can enter a shell command to be run.
You can return to "read" mode without running a command
by entering an EOF (control-D) at the ":" prompt.
The ":!" and ":shell" commands are equivalent to this.
.IP "-" 1i
Brings up the "-" prompt so you can type the number of old message lines
you would like to see redisplayed.
The redisplay can be interrupted to return you quickly to "read" mode,
unless you have an output filter defined, in which case interrupting
a tailback exits you from party.
.IP ":" 1i
Brings up the ":" prompt so you can type in a special command.
Special commands are listed below.
You can return to "read" mode without entering any command
by entering an EOF (control-D) at the ":" prompt.
.IP "/" 1i
Brings up the "/" prompt for making vaguely MUD-like sound effects.
For a list of possible noises (if any), do a "/" with no argument.
For more information about noises, see the "makenoise" option below.
.IP "=" 1i
Prints all current settings of user options.
This is equivalent to ":print".
.IP "? or h" 1i
Prints the help message.
The help message can be interrupted to return you quickly to "read"
mode.
The "h" command goes away if you use the firstchar option.
.IP "q or EOF" 1i
Exits party.  An interrupt will also work.
The "q" command goes away if you use the firstchar option.
.LP
The following special commands are supported at the colon prompt.
All commands can be abbreviated.
.IP ":join <channel>" 1i
This exits from the current channel and joins the named channel.
The '#' command is a shorthand for this.
.IP ":who [-cnt]" 1i
This lists all users currently in party, together with the time they
entered and their current channel numbers.  A flag may be given to indicate
whether to sort the names by channel (-c, the default), by name (-n), or
by start time (-t).
.IP ":list <channel>" 1i
This lists all current channels telling how many users are in each.
The channel you are in is marked with a carat.
.IP ":name <newname>" 1i
This changes your name in the current channel to the newname specified.
It only works in channels where the "rename" option is enabled.
.IP ":read <filename>" 1i
This reads the lines of the named file into the party file.
The maximum number of lines that can be read into the file is set by
"readlim" option.
.IP ":save [-<count>] <filename>" 1i
This saves a copy of the log for the current party channel in the named
file.  If a count is given, only the last
.I <count>
lines of the log file are saved.
.IP ":set <option-list>" 1i
This sets the options listed.
.IP ":print <option-list>" 1i
This displays the current values of the options listed.
If no options are given, a selection of the more interesting options are
displayed
(this is equivalent to the "=" command).
If the argument "all" is given, then all options, including system options
are displayed.
.IP ":shell <command>" 1i
This executes the a shell escape.  You can also just use the "!" command
or the ":!" command.
The shell used to execute the command is determined by the "shell" and
"fastshell" options described below.
.IP ":close" 1i
Closes the current channel.  This means not permitting any users not already
in the channel to join, and depermitting the log file so no outsider users
can read it.  This only works if the "mayclose" option is enabled.
In some installations, this may not exist at all, or it may not
depermit the log file.  Note that things typed in a closed channel become
readable to everyone again if the channel is opened.
The only way to get anything like real secrecy is to close a nokeeplog channel
and never reopen it.  Just let it be destroyed when all users leave.
.IP ":open" 1in
Opens the current channel, if it has been closed.
Opening a channel makes the whole history of the conversation there readable
to all.
.IP ":invite <login>" 1in
Add the given login id to the list of users who may join the current channel,
if it is closed.
This does not send any kind of message to that user.
.IP ":ignore <user>..." 1i
The named users will be added to the list of ignored users.
No messages from ignored users will be displayed.
If no user names are given on the command line,
the names of all currently ignored users will be listed.
Note that if you are in a channel where name changing is allowed and
.I uniquename
is set, then if a person you are ignoring changes his name,
your ignore list will automatically be updated to the new name.
.IP ":notice <user>..." 1i
The named users will be deleted from the list of ignored users,
so that their messages will once again be displayed.
If no user names are given on the command line,
all currently ignored users will be removed from the list.
.IP ":back <count>" 1i
Backs up in the file by the given number of lines.
The '-' command character is a shorthand for this.
.IP ":version" 1i
Prints the current party version number.
.IP ":quit" 1i
Yet another way to exit party.
.LP
The behavior of
.I party
can be altered with a number of options.
There are six ways to set party options.
Options in option lists are always separated by spaces.
.IP "1)" 0.5i
Compiled in defaults.
These should be set to sane values by whoever installed
.I party
on your system.
.IP "2)" 0.5i
Partytab file.
The file /usr/bbs/party/partytab contains several lines of the form:
.IP "" 1i
<program-name> <option-list>
.IP "" 0.5i
When invoked, party searches the partytab file for a line with the
program-name that it was invoked under.
(For example, if there was a link to the party program named
.I cb
and you ran the 
.I cb
program, it would look for a line starting with
. I cb
in the partytab).
It then applies the option settings listed after that program-name.
This is an easy way to change the default options without recompiling,
and allows several "different" versions of the
.I party
program to be
supported on the system without having to make separate physical copies.
.IP "3)" 0.5i
PARTYOPTS environment variable.
This may be set to a list of options which will always be set when you
enter any
.I party
program.
.IP "4)" 0.5i
Command line.
A list of options may be given as arguments on the command line of any
party program.
.IP "5)" 0.5i
Chantab file.
The file /usr/bbs/party/chantab contains several lines of the form:
.IP "" 0.5i
<channel-name-pattern> <option-list>
.IP "" 0.5i
When you enter a party channel,
party searches for a chantab line starting with a pattern that matches the
channel name.  The syntax of the pattern is similar to the shell's filename
substitution.
For example, a ? matches any one character, a * matches any sequence of
characters.
Party sets the options listed after the first pattern that matches
the channel name.
This allows different channels to have different rules.
Normally only system options should be set in the chantab.
Most of these are automatically reset to the state they were in
after reading the partytab every time you can change channels.
Channel names that don't match any pattern in the chantab file may not be used.
.IP "6)" 0.5i
The ":set" command.
Options may be reset while the program is running by using
the "set" special command at the colon prompt.
.LP
The options are read in the order listed above,
so options set from the last places listed
override ones set from the earlier places.
.LP
There are four types of values an option named <option> can take.
Boolean values are set with "<option>" and reset with "no<option>"
A few options can also take take prefixes other than "no".
Numeric options are set with "<option>=<value>".
String options are set with "<option>=<string>".
If the string includes spaces or tabs,
it must be quoted with either single or double quotes.
.LP
Most options are user-settable, but others are reserved for the administrator
to use in configuring the party program and individual channels.
Those can be set only in the partytab or chantab files.
The user-settable options are:
.IP "alias=<name>"
This is the default name to use when joining a channel where the
rename option is enabled.
It defaults to the user's real login name.
.IP "[no]arg" 1i
If noarg is set,
.I party
does not read options from the command argument list.
Obviously this is only meaningful in the partytab or in PARTYOPTS,
since otherwise they has already been read.
The default is arg.
.IP "back=<count>" 1i
This defines the number of old lines to display when you enter
.I party
or change channels.
The default is back=10.
.IP "[no][see]bs" 1i
This determines how to handle backspaces in messages.
The default, bs, is just to print them.
This allows tricky users to "cursor-dance", possibly backing over their
names and changing them.
Setting nobs causes all backspace characters to be stripped out.
Setting seebs causes backspace characters to be displayed as "^H".
noseebs is the same as bs.
.IP "[no]colon" 1i
If nocolon is set,
.I party
does not recognize the ":" command to set options.
This may be used, together with noenv and noarg, to create a party program
that cannot be customized by the user.
The default is colon.
.IP "[no][see]control" 1i
This determines how to handle other control characters in messages.
The default, control, is just to print them.
Setting nocontrol causes all control characters to be stripped out.
Setting seecontrol causes control characters to be displayed as "^E"
or whatever.
.IP "[no]env" 1i
If noenv is set,
.I party
does not read options from the PARTYOPTS variable.
Obviously this is only meaningful in the partytab,
since otherwise they has already been read.
The default is env.
.IP "knockwait=<seconds>"
This tells how long to wait for an invitation when attempting to join a
closed channel.
The default is 30 seconds.
The user can always interrupt the wait if he gets impatient.
.IP "[no]fastshell" 1i
If fastshell is set, party will execute most shell escape commands directly
instead of starting a shell to execute them.
It will start a shell for commands that look like they contain
wildcards, IO-redirection, or other things that normally require
a shell to execute.
This will generally speed up the execution of shell escapes.
If nofastshell is set, all commands will actually be processed by the shell
in the shell variable.
Mainly nofastshell is useful if you have weird shells or if you have
shell=/bin/csh and want aliases defined in your .cshrc file to work.
The default is fastshell.
.IP "filter=<cmd>" 1i
This starts the named command as an output filter.
Only messages and entry/exit banners are printed through the filter.
Prompts, help messages, and shell escape output are not.
Output filters can be used in many clever ways.
For example, to stop displaying messages from the user 'janc',
you could set filter="grep -v '^janc:'".
filter commands are processed by the shell specified by the shell option,
so pipes and aliases and such like can be used.
If there is already a filter defined, setting a new one will turn off
the old one.
The default filter is filter="" (ie, there is no default).
.IP "[no]filter" 1i
If you have a filter defined, this turns it on and off.
Note that setting a new filter automatically turns it on.
.IP "[no]firstchar" 1i
If firstchar is enabled, then whatever key you hit to bring up the
">" input prompt, will also become the first character of your input text.
This is mainly meant to be less confusing for regular IRC users.
In this mode, party does all the input processing instead of letting the
Unix tty driver do it.
This means there are some inevitable differences in the way input is processed,
but most are minor.
As a side effect, the "q" command to quit and the "h" command to get help
go away, since otherwise you wouldn't easily be able to enter lines starting
with these letters.
You can still quit with and EOF character or a ":q" command.
You can still get help with w "?" character.
The default is nofirstchar.
.IP "fullmesg=<string>" 1i
This is a text string to be printed out when a user attempts to join party
when the
.I capacity
option is set and party is full.

contrib/party/party-2.13/party.c
/*      Party - (c) July 1990 - Jan Wolter
 *
 *	The party program simulates a party line, where any number of users
 *	can talk at the same time.  It was originally developed under System
 *	III Unix, and as such communicates through a log file rather than
 *	using interprocess communications.  It can only be used between
 *	different computers if they can share the log file via some sort
 *	of NFS-like mechanism.  It should however work under most any version
 *	of Unix.
 *	
 *      This program may be freely distributed without permission of the
 *	author.  No permission is needed to run this software anywhere.
 *	However selling the program itself would, of course, be inappropriate.
 *
 *						Dr. Jan Wolter
 *						janc@unixpapa.com
 * 
 *  RELEASE HISTORY (version numbers before 2.4 were assigned retroactively)
 *
 *  version 1.0:   Original two-process party program by Marcus Watts.
 *  version 1.8:   Jan Wolter's bug fixes, changes in output format, added
 *                 partytab
 *  version 2.0:   Single process rewrite by Jan Wolter.
 *  version 2.1:   Added user settable options.
 *  version 2.2:   Added multiple channels, external filter control.
 *  version 2.3:   Major internal clean-up, reorganized output filters, special
 *	           commands added, noises added.
 *  version 2.4:   Added "who" command, BSD compatibility.  The Grex release.
 *  version 2.5:   Added volatile channels.  :name command.
 *  version 2.6:   Minor bug fixes, System V and SCO portability patches.
 *  version 2.7:   Closable channels, BSDI bug fixes, other minor bug fixes.
 *  version 2.8:   Login names more than 8 characters for next BSDI version;
 *                 norepeat option added; sorted :who listings; showevent.
 *  version 2.9:   Idle time outs, capacity limits, intro commands, no control
 *                 characters in shelled command name
 *  version 2.10:  Add "raw" option and :ignore and :notice commands.
 *  version 2.10a: SIGTERM makes a clean exit.  Fixed Marcus's SIGPIPE bug.
 *                 :list in three columns.  Added "uniquename".  Better ttyname
 *                 checking.  Use MAIL environment variable.  Fixed /dev/tty
 *                 cloaking bug.  :ignore tracks name changes.
 *  version 2.10b: fix finduser's overflow of logname variable.
 *  version 2.10d: fix control characters in user names via PARTYOPTS.
 *  version 2.11:  use select() if NOSELECT is not defined.
 *  version 2.11a: turn NOSELECT off - it doesn't work and never will!
 *  version 2.12:  Fixed signals for Linux.  ANSIfication.  AUTOCONFication.
 *  version 2.13:  Default shell from environment.  Tilde expansion on :read
 *                 and :save.
 */

#include "party.h"
#include "opt.h"
#include <sys/stat.h>
#include <errno.h>
#include <ctype.h>
#include <setjmp.h>
#include <pwd.h>

char *version= "2.13";		/* current party version */

int inhelp=0;			/* Are we printing help text? */
off_t tailed_from=0L;		/* File offset we last tailed back from */
FILE *wfd;			/* Party file open for write */
int rst;			/* Party file open for read */
char *channel= NULL;		/* Current channel name (NULL is outside)*/
char *progname;			/* Program name */
char mailfile[80];		/* Name of mail file */
int  mailfuse;			/* counter for deciding when to check mail */
long mailsize;			/* old size of mail file */
uid_t real_id, eff_id;		/* My real and effective group or user id */
char inbuf[BFSZ+INDENT+2];      /* Text buffer - first 10 for "name:   " */
char *txbuf= inbuf + INDENT;    /* Text buffer - pointer to respose portion */
				/*               of inbuf */
FILE *debug= NULL;		/* Debug file, if one is open */
RETSIGTYPE (*oldsigpipe)();	/* Old SIGPIPE handler */

jump_buf jenv;

int
main(int argc, char **argv)
{
    register int n;
    char ch, *pnl;
    time_t lastaction;
#ifndef NOSELECT
    fd_set input_set, iset;
#endif

    progname= leafname(argv[0]);	/* Get the program name */

    /* Look self up in utmp */
    if (finduser(logtty,logname,&logtime)) exit(1);

    setrealname();		/* Get the user name */
    initopts();			/* Set default options */
    readtab();			/* Read options from partytab */
    if (who_open()) exit(1);	/* Open the partytmp file */

    /* Check if we are just doing a pwho command */
    if (opt[OPT_USERLIST].yes)
    {
	who_list(stdout,'c');
	exit(0);
    }

    /* Check if there is room for one more in party - race conditions
     * could allow the capacity to be exceeded if two people enter at
     * once, but who cares?
     */
    if (opt[OPT_CAPACITY].yes &&
	who_count() >= atoi(opt[OPT_CAPACITY].str))
    {
	printexec(stderr,opt[OPT_FULLMESG].str);
	exit(1);
    }

    /* If possible, set OPT_COLS from TERMCAP at this point */
    setcols();

    /* Read options from PARTYOPTS env var */
    if (opt[OPT_ENV].yes && (pnl= getenv("PARTYOPTS")))
	parseopts(pnl,0);

    /* Read options from parameters */
    if (opt[OPT_ARG].yes)
	for (n= 1; n < argc; n++)
	    parseopts(argv[n],2);

#ifdef SUID
    real_id= getuid(); eff_id= geteuid();
#endif
#ifdef SGID
    real_id= getgid(); eff_id= getegid();
#endif

    initmodes();
    setmailfile();
    initoutput();
    if (who_enter()) done(1);

    printexec(stderr,opt[OPT_INTRO].str);

    /* Put up the signal handlers */
    signal(SIGHUP,(RETSIGTYPE (*)())hup);
    signal(SIGQUIT,(RETSIGTYPE (*)())intr);
    signal(SIGINT,(RETSIGTYPE (*)())intr);
    signal(SIGTERM,(RETSIGTYPE (*)())term);
    oldsigpipe= signal(SIGPIPE,SIG_IGN);
#ifdef NOSELECT
    signal(SIGALRM,(RETSIGTYPE (*)())alrm);
#endif /*NOSELECT*/
#ifdef SIGTSTP
    signal(SIGTSTP,(RETSIGTYPE (*)())susp);
#endif /*SIGTSTP*/
#ifdef WINDOWS
    signal(SIGWINCH,(RETSIGTYPE (*)())setcols);
#endif /*WINDOWS*/

    if (join_party(opt[OPT_START].str)) exit(1);

    /* Get in cbreak/noecho mode */
    STTY(0,&cbreak);

#ifndef NOSELECT
    /* Set up the input file set for the select call */
    FD_ZERO(&input_set);
    FD_SET(0,&input_set);
    FD_SET(rst,&input_set);
#endif

    /* Main loop */
    lastaction= time((time_t *)0);
    for(;;)
    {
#ifdef NOSELECT
	/* Copy text while I can -- otherwise go to input mode */
	if (output())
	{ /*}*/
	    /* Wait four seconds for a command */
	    alarm(4);
	    if (!setjump(jenv,1))
	    {
		read(0,&ch,1);	/* Alarm busts us out of here */
		alarm(0);
		docmd(ch);
		lastaction= time((time_t *)0);
	    }
#else
	/* Wait for input in file or from user */
	iset= input_set;
	select(rst+1, &iset, NULL, NULL, NULL);

	/* process typed command from user */
	if (FD_ISSET(0,&iset))
	{
	    read(0,&ch,1);
	    docmd(ch);
	    lastaction= time((time_t *)0);
	}

	/* process new text in file */
	if (FD_ISSET(rst,&iset))
	{
	    while (!output())
		;
#endif /*NOSELECT*/

	    /* Check for new mail or idleness */
	    if (mailfuse-- == 0)
	    {
		struct stat mailstat;

		/* Check if he has been idle a long time */
		if (opt[OPT_IDLEOUT].yes &&
		    time((time_t *)0) - lastaction >
		       atoi(opt[OPT_IDLEOUT].str)*60)
		{
		    fprintf(stderr,
		      "You have been idle in party for more than %s minutes\nBye!\n",
		      opt[OPT_IDLEOUT].str);
		    done(0);
		}

		/* Check if size of mailfile has increased */
		mailfuse= MAILDELAY;
		if (stat(mailfile,&mailstat))
		    mailsize= 0L;
		else
		{
		    if (mailstat.st_size > mailsize)
			fprintf(stderr,"You have more mail\n");
		    mailsize= mailstat.st_size;
		}
	    }
	}
    }
}


#ifdef NOSELECT
RETSIGTYPE alrm()
{
    signal(SIGALRM,(RETSIGTYPE (*)())alrm);
    longjump(jenv,1);
}
#endif /*NOSELECT*/


/* HELP: Prints a file name to the screen.  An interrupt brings you back to
 * party.  If complain is false, it silently does nothing when the file
 * doesn't exist.
 */

void help(char *filename, int complain)
{
    register int hf;
    register int n;

    if ((hf= open(filename,0)) < 0)
    {
	if (complain) err("Cannot open file %s\n",filename);
    }
    else
    {
	if (!setjump(jenv,1))
	{
	    inhelp= 1;
	    while((n= read(hf,txbuf,BFSZ)) > 0)
		write(1,txbuf,n);
	}
	close(hf);
	inhelp= 0;
    }
}

/* EXPTILDE:  given a file name that starts with a ~, return a copy in malloc'ed
 * memory that has the the ~ expanded.  If there is no ~ or the ~ cannot be
 * expanded return NULL.
 */

char *exptilde(char *filename)
{
    char *dir= NULL;
    char *slash;
    struct passwd *pw;

    /* Need a ~ and a slash to get started */
    if (filename[0] != '~' || (slash= strchr(filename+1,'/')) == NULL)
	return NULL;

    /* Figure out the home directory */
    *slash= '\0';
    if (slash == filename + 1)
    {
	if ((dir= getenv("HOME")) == NULL)
	{
	    if ((pw= getpwuid(getuid())) != NULL)
		dir= pw->pw_dir;
	}
    }
    else if ((pw= getpwnam(filename+1)) != NULL)
	dir= pw->pw_dir;
    *slash= '/';

    if (dir == NULL)
	return NULL;

    /* Construct the path */
    filename= (char *)malloc(strlen(dir) + 1 + strlen(slash+1));
    sprintf(filename,"%s/%s",dir,slash+1);

    return filename;
}


/* READFILE: This reads a file into the partylog, much as if you had typed it
 * in, except that each line is prepended with a space instead of your logname.
 * Files must be readable *both* to party and to the user. 
 */

void readfile(char *filename)
{
    FILE *fp;
    int readlim= convert(opt[OPT_READLIM].str);
    int lines= 0;
    char *tilde;

    if (readlim == 0)
    {
	err("File reading not enabled in this channel\n");
	return;
    }

    /* Do tilde expansion */
    if ((tilde= exptilde(filename)) != NULL)
	filename= tilde;

    /* Can only open files if user and party can read it */
    be_user();
    if (access(filename,4) || (fp= fopen(filename,"r")) == NULL)
    {
	err("Cannot open input file %s\n",filename);
	be_party();
	return;
    }
    if (tilde != NULL) free(filename);
    be_party();

    /* Print user's name */
    txbuf[0]= '\n';
    txbuf[1]= '\0';
    LOCK(wfd);
    fseek(wfd,0L,2);
    fputs(inbuf,wfd);
    fflush(wfd);
    
    txbuf[0]= ' ';
    while ((lines++ < readlim)  && fgets(txbuf+1,BFSZ-1,fp))
    {
	fseek(wfd,0L,2);
	fputs(txbuf,wfd);
	if (strchr(txbuf,'\n') == NULL) fputc('\n',wfd);
	fflush(wfd);
    }
    UNLOCK(wfd);
    fclose(fp);
}


/* CHN_FILE_NAME:  return the channel file name for the current channel.
 * Somebody needs to free this.
 */

char *chn_file_name(char *chn, int keeplog)
{
    char *file;

    file= (char *)malloc(strlen(opt[OPT_DIR].str)+CHN_LEN+6);
    sprintf(file,"%s/%s.%s",
	opt[OPT_DIR].str,
	chn,
	keeplog?"log":"tmp");
    return(file);
}


/* JOIN_PARTY: changes from the current channel to the new channel.  Either one
 * of those could be NULL.  Returns 1 if the named channel doesn't exist.
 */

int join_party(char *nch)
{
    char *file;
    FILE *tmp_wfd;
    int tmp_rst,oumask,waskept;
    time_t now= time((time_t *)0);
    char newchannel[CHN_LEN+1];
    static char ch[CHN_LEN+1];

    if (debug) db("join_party %s\n",nch?nch:"(nil)");

    waskept= opt[OPT_KEEPLOG].yes;

    /* If we are to enter a new channel, make sure we can */
    if (nch != NULL)
    {
	/* Make a local copy of name */
	strncpy(newchannel,nch,CHN_LEN);
	newchannel[CHN_LEN]= '\0';

	if (badname(newchannel))
	{
	    err("improper channel name\n");
	    return(1);
	}

#ifndef NOCLOSE
	if (!enter_closed(newchannel))
	    return(1);
#endif /*NOCLOSE*/

	/* Set the channel options */
	if (chnopt(newchannel))
	{
	    err("channel %s does not exist\n",newchannel);
	    return(1);
	}
	if (debug) printopts(debug,0,' ',"all");

	/* Construct the channel log file name */
	file= chn_file_name(newchannel,opt[OPT_KEEPLOG].yes);

	/* Open the partyfile to read - if it doesn't exist create it */
	if ((!opt[OPT_MAYCLOSE].yes && access(file,4)) ||
	    (tmp_rst= open(file,O_RDONLY)) < 0)
	{
	    /* I don't have read access, or the file doesn't exist */
	    oumask= umask(000);
	    if (!access(file,0) ||
	       (tmp_rst= open(file,O_RDONLY|O_CREAT,
		    opt[OPT_MAYCLOSE].yes ? DEP_MODE : CHN_MODE)) < 0)
	    {
		err("channel %s not accessible\n(%s unreadable)\n",
		    newchannel,file);
		free(file);
		umask(oumask);
		if (channel != NULL) chnopt(channel);
		return(1);
	    }
	    umask(oumask);
	}

	/* Open partyfile to write */
	if ((tmp_wfd= fopen(file,"a")) == NULL)
	{
	    err("Cannot write partyfile %s\n",file);
	    close(tmp_rst);
	    free(file);
	    if (channel != NULL) chnopt(channel);
	    return(1);
	}
	free(file);

	if (debug) db("join_party: new channel open\n");
    }

    /* If we were in some channel, get out */
    if (channel != NULL)
    {
	    if (nch == NULL)
		    sprintf(txbuf,"---- %s leaving (%.12s)\n",
			    name,ctime(&now)+4);
	    else
		    sprintf(txbuf,
			    "---- %s switching to channel %s (%.12s)\n",
			    name,newchannel,ctime(&now)+4);
	    write(out_fd,txbuf,strlen(txbuf));

	    /* Put departure message in the file */
	    append(txbuf,wfd);

	    /* Close the old files */
	    fclose(wfd);
	    close(rst);

	    /* If it was a temporary channel and it is empty, delete it */
	    if (!waskept && who_empty(channel))
	    {
		file= chn_file_name(channel,0);
		unlink(file);
		free(file);
#ifndef NOCLOSE
		file= usr_file_name(channel);
		unlink(file);
		free(file);
#endif /*NOCLOSE*/
	    }
	if (debug) db("join_party: old channel closed\n");
    }

    /* If we are to enter a new channel, finish the job */
    if (nch != NULL)
    {
	/* Make the new file the current file */
	rst= tmp_rst;
	wfd= tmp_wfd;
	tailed_from= 0L;

	/* Set streams to close on exec */
	fcntl(rst,F_SETFD,1);
	fcntl(fileno(wfd),F_SETFD,1);

	/* Print Channel Banner */
	if (channel != NULL)
	{
	    if (opt[OPT_CHANINTRO].yes)
		printexec(stderr,opt[OPT_CHANINTRO].str);
	    else
		fprintf(stderr,
		"\n================== channel %s ==================\n",
		newchannel);
	    fprintf(stderr,"Options:");
	    if (printopts(stderr,8,' ',"chan"))
		fprintf(stderr,"normal\n\n");
	    else
		fputc('\n',stderr);
	}

	/* Get the user's name or pseudonym */
	setname(newchannel);

	/* Put in a join message */
	LOCK(wfd);
	fseek(wfd,0L,2);
	if (channel == NULL)
	    fprintf(wfd,"---- %s joining (%.12s)\n",
		    name,ctime(&now)+4);
	else
	    fprintf(wfd,"---- %s joining from channel %s (%.12s)\n",
		    name,channel,ctime(&now)+4);
	fflush(wfd);
	UNLOCK(wfd);

	/* Wind back a few lines */
	lseek(rst,0L,2);			/* Goto end of file */
	backup(convert(opt[OPT_BACK].str));

	strcpy(channel=ch,newchannel);

	who_chan();
    }
    else
	who_exit();

    return(0);
}


/* SETMAILFILE:  Start monitering the user's mail directory so we can notify
 * him when he has new mail.
 */

void setmailfile()
{
    struct stat mailstat;
    char *env;

    if ((env= getenv("MAIL")) != NULL)
	strcpy(mailfile,env);
    else
    {
	strcpy(mailfile,opt[OPT_MAILDIR].str);
	strcat(mailfile,"/");
	strcat(mailfile,realname);
    }
    if (stat(mailfile,&mailstat))
	    mailsize= 0L;
    else
	    mailsize= mailstat.st_size;
    mailfuse= MAILDELAY;
}

/* LEAFNAME: Return the trailing name from a pathname.
 */

char *leafname(char *c)
{
    register char *t;

    if ((t= strrchr(c,'/')) == NULL)
	return(c);
    else
	return(t+1);
}


/* BADNAME:  Check a channel name to make sure it contains only legal
 * characters: namely printable, non-space, non-dot, non-slash characters.  We
 * could probably stand dots, but we don't.
 */

int badname(char *chn)
{
    char *badchr= " ./#";

    for(;*chn;chn++)
	if (!isascii(*chn) || !isprint(*chn) || strchr(badchr,*chn))
	    return(1);
    return(0);
}

/* CONVERT:  Convert a string to a non-negative integer.  This does a bit more
 * checking than atoi().  It returns -1 if the string is not syntactially
 * correct.
 */

int convert(char *c)
{
    register int n=0;

    /* Skip leading spaces and tabs */
    while (*c == ' ' || *c == '\t')
	c++;

    for ( ; *c != '\n' && *c != '\0'; c++)
    {
	if (*c >= '0' && *c <= '9')
	    n= n*10 + *c - '0';
	else
	    return (-1);
    }
    return(n);
}

#ifdef LOCK_FCNTL
/* Helper function to do fcntl locks.
 */

void setlock(int fd, int type)
{
    struct flock lk;

    lk.l_type= type;
    lk.l_whence= 0;
    lk.l_start= 0L;
    lk.l_len= 0L;
    fcntl(fd,F_SETLKW,&lk);
}
#endif /*LOCK_FCNTL*/

#ifdef LOCK_LOCKING
/* Keep calling locking() until it works or until we get sick and tired of
 * calling it.  This apparantly needs to be done because it doesn't block
 * properly. There may be problems with this.
 */

void chk_lock(FILE *file, int request)
{
    register int i= 5;

    while (locking(fileno(file), request, -1L) && i--)
	sleep(1);
}
#endif /*LOCK_LOCKING*/


/* BACKUP:  This seeks backwards by the given number of lines.
 */

off_t backup(int lines)
{
#define BUB_SIZE 512
    char bub[BUB_SIZE];
    register char *p;
    off_t off;
    int n;

    if (lines <= 0) return 0;

    p= bub;
    off= lseek(rst,0L,1);

    /* Scan backwards, counting newlines */
    do
	if (p-- == bub)
	{
	    if (off == 0L) break;
	    /* Fetch a new block of data */
	    off -= (n= (off > BUB_SIZE) ? BUB_SIZE : off);
	    lseek(rst, off, 0);
	    read(rst, bub, n);
	    p= bub + n - 1;
	}
    while ( *p != '\n' || lines-- > 0);

    /* Position file pointer */
    return(lseek(rst, off+(p-bub)+1, 0));
}


/* INTR: Handle user interrupts.  If he is tailing back or printing a help
 * file, just interrupt that.
 */

RETSIGTYPE intr()
{
    if (inhelp)
    {
	/* Old Unixes lost interupt handlers after interupts.  Modern
	 * Unixes don't need this but we do it anyway, */
	signal(SIGQUIT,(RETSIGTYPE (*)())intr);
	signal(SIGINT,(RETSIGTYPE (*)())intr);
	longjump(jenv,1);
    }
    if (tailed_from > lseek(rst,0L,1))
    {
	signal(SIGQUIT,(RETSIGTYPE (*)())intr);
	signal(SIGINT,(RETSIGTYPE (*)())intr);
	printf("\nTailback Interrupted...\n");
	lseek(rst,tailed_from,0);
	return;
    }

#ifdef NOSELECT
    alarm(0);
#endif /*NOSELECT*/

    signal(SIGINT,SIG_IGN);
    signal(SIGQUIT,SIG_IGN);

    join_party(NULL);

    STTY(0,&cooked);
    kill_filter();
    exit(0);
}


/* TERM:  Exit the program because we got a SIGTERM. This is like a hup(), but
 * prints a message to the user. */

RETSIGTYPE term()
{
    if (debug) db("killed\n");

#ifdef NOSELECT
    alarm(0);
#endif /*NOSELECT*/
    signal(SIGINT,SIG_IGN);
    signal(SIGQUIT,SIG_IGN);

    join_party(NULL);

    kill_filter();
    STTY(0,&cooked);

    printf("\nParty Process Killed.\n");

    exit(0);
}


/* DONE:  Exit the program at the user's command.  This should be used just
 * about everywhere you would normally call exit.
 */

void done(int rc)
{
    if (debug) db("exiting\n");

#ifdef NOSELECT
    alarm(0);
#endif /*NOSELECT*/

    signal(SIGINT,SIG_IGN);
    signal(SIGQUIT,SIG_IGN);

    join_party(NULL);
    stop_filter();
    STTY(0,&cooked);
    exit(rc);
}

/* HUP:  Handle Hangups.  The only difference betwen this and done() is that
 * we are more brutal about killing off the filter.
 */

RETSIGTYPE hup()
{
    if (debug) db("hangup\n");

#ifdef NOSELECT
    alarm(0);
#endif /*NOSELECT*/
    signal(SIGINT,SIG_IGN);
    signal(SIGQUIT,SIG_IGN);

    join_party(NULL);

    kill_filter();
    STTY(0,&cooked);
    exit(0);
}

/* SUSP: Suspend execution temporarily.
 */

#ifdef SIGTSTP
RETSIGTYPE susp()
{
#ifdef F_TERMIOS
    struct termios old;
#endif
#ifdef F_STTY
    struct sgttyb old;
#endif
    int mask;
    int was_shelled;

    if (debug) db("suspended\n");

    if (!(was_shelled= who_isout())) who_shout("^");
#ifdef NOSELECT
    alarm(0);
#endif /*NOSELECT*/
    mask= sigblock(sigmask(SIGTSTP));

    GTTY(0,&old);
    STTY(0,&cooked);

    signal(SIGTSTP,SIG_DFL);
    sigsetmask(0);
    kill(getpid(),SIGTSTP);

    /* STOP HERE */

    sigsetmask(mask);
    signal(SIGTSTP,(void (*)())susp);

    initmodes();
    STTY(0,&old);
    if (!was_shelled) who_shin();

    if (debug) db("restarted\n");
}
#endif /*SIGTSTP*/

void db(char *msg, ...)
{
    va_list ap;
    time_t now;

    if (debug)
    {
        va_start(ap, msg);
	now= time((time_t *)0);
        fprintf(debug,"%8.8s:  ",ctime(&now)+11);
        vfprintf(debug,msg,ap);
	fflush(debug);
	va_end(ap);
    }
}


void err(char *msg, ...)
{
    va_list ap;
    time_t now;

    va_start(ap, msg);
    fprintf(stderr,"%s error: ",progname);
    vfprintf(stderr,msg,ap);

    if (debug)
    {

	now= time((time_t *)0);
	fprintf(debug,"%8.8s:  ",ctime(&now)+11);
	vfprintf(debug,msg,ap);
	fflush(debug);
    }
    va_end(ap);
}
